File: src/install_page/install_page.py
from gi.repository import Adw, Gtk, GLib
from .host_info import HostInfo
from .select_page import SelectPage
from .pending_page import PendingPage
from .sidebar_button import SidebarButton
from .loading_status import LoadingStatus
from .package_install_worker import PackageInstallWorker
from .error_toast import ErrorToast

@Gtk.Template(resource_path="/io/github/flattool/Warehouse/install_page/install_page.ui")
class InstallPage(Adw.BreakpointBin):
	__gtype_name__ = "InstallPage"
	gtc = Gtk.Template.Child
	
	break_point = gtc()
	split_view = gtc()
	multi_view = gtc()
	select_page = gtc()
	pending_page = gtc()
	status_stack = gtc()
	loading_view = gtc()
	installing_view = gtc()
	bottom_sheet = gtc()
	bottom_child = gtc()
	bottom_label = gtc()
	
	# Referred to in the main window
	#	It is used to determine if a new page should be made or not
	#	This must be set to the created object from within the class's __init__ method
	instance = None
	page_name = "install"
	
	current_installation = ""
	current_remote = None
	did_error = False
	
	def start_loading(self):
		self.total_added_packages = 0
		self.bottom_bar_visual_handler(False)
		self.status_stack.set_visible_child(self.loading_view)
		self.select_page.start_loading()
		self.pending_page.reset()
		
	def end_loading(self):
		self.select_page.end_loading()
		self.status_stack.set_visible_child(self.multi_view)
		
	def install_callback(self):
		HostInfo.main_window.refresh_handler()
		if not self.did_error:
			HostInfo.main_window.toast_overlay.add_toast(Adw.Toast(title=_("Installed Packages")))
			
	def install_error_callback(self, user_facing_label, error_message):
		self.did_error = True
		GLib.idle_add(lambda *_: HostInfo.main_window.toast_overlay.add_toast(ErrorToast(user_facing_label, error_message).toast))
		
	def install_packages(self, package_requests):
		self.did_error = False
		if PackageInstallWorker.install(package_requests, self.installing_status, self.install_callback, self.install_error_callback):
			self.status_stack.set_visible_child(self.installing_view)
			
	def bottom_bar_visual_handler(self, is_added):
		total = self.total_added_packages
		if total == 0:
			self.bottom_child.set_reveal_child(False)
			self.bottom_sheet.set_bottom_bar(None)
		elif total == 1:
			self.bottom_label.set_label(_("{} Pending Package").format(1))
			if is_added:
				self.bottom_sheet.set_bottom_bar(self.bottom_child)
				self.bottom_child.set_reveal_child(True)
		else:
			self.bottom_label.set_label(_("{} Pending Packages").format(total))
			
	def package_added(self):
		self.total_added_packages += 1
		self.bottom_bar_visual_handler(True)
		
	def package_removed(self):
		self.total_added_packages -= 1
		self.bottom_bar_visual_handler(False)
		
	def __init__(self, main_window, **kwargs):
		super().__init__(**kwargs)
		self.instance = self
		
		# Extra Object Creation
		self.installing_status = LoadingStatus(_("Installing Packages"), _("This could take a while"), True, PackageInstallWorker.cancel)
		self.total_added_packages = 0
		
		# Connections
		
		# Apply
		self.select_page.results_page.pending_page = self.pending_page
		self.select_page.results_page.install_page = self
		self.loading_view.set_content(LoadingStatus(_("Loading Installation Options"), _("This should only take a moment")))
		self.installing_view.set_content(self.installing_status)

File: src/install_page/install_page.ts
import GObject from "gi://GObject?version=2.0"
import Gio from "gi://Gio?version=2.0"
import Gdk from "gi://Gdk?version=4.0"
import Adw from "gi://Adw?version=1"
import Gtk from "gi://Gtk?version=4.0"

import { from, GClass, Child, Property, Signal, SimplerAction } from "../gobjectify/gobjectify.js"
import { BasePage } from "../widgets/base_page.js"
import { SelectPage } from "./select_page.js"
import { PendingPage } from "./pending_page.js"

import "../widgets/sidebar_button.js"

type PageNamges = "installing-page" | "loading-page" | "content-page"

@GClass({ template: "resource:///io/github/flattool/Warehouse/install_page/install_page.ui" })
export class InstallPage extends from(BasePage, {
	installing: Property.readwrite.bool(),
	_select_page: Child<SelectPage>(),
	_pending_page: Child<PendingPage>(),
}) {
	constructor(params?: typeof InstallPage.$params) {
		params ??= {}
		params.icon_name = "warehouse:arrow-pointing-at-line-down-symbolic"
		params.sidebar_title = _("Install Packages")
		super(params)
		this._select_page.$connect("queue-add", (__, row, remotes) => {
			const remote = remotes?.[0]
			const result = row?.result
			const installation = result?.installation
			if (!result || !remote || !installation) return
			this._pending_page.add(result, remote)
			if (this._pending_page.has_result_by_id(installation, result.application)) {
				row.kind = "added"
			}
		})
		this._select_page.$connect("row-created", (__, row) => {
			const result = row?.result
			const installation = result?.installation
			if (!row || !result || !installation) return
			if (this._pending_page.has_result_by_id(installation, result.application)) {
				row.kind = "added"
			}
		})
		this._pending_page.$connect("queue-remove", (__, result) => {
			const inst_name = result?.installation?.name
			const app = result?.application
			if (!inst_name || !app) return
			this._select_page.result_deslected(inst_name, app)
		})
	}

	// TODO: reset select and pending pages when the loading prop changes

	protected _get_visible_page(__: this, loading: boolean, installing: boolean): PageNamges {
		if (loading) return "loading-page"
		if (installing) return "installing-page"
		return "content-page"
	}

	protected _get_reveal_bottom_bar(__: this, n_pending: number): boolean {
		return n_pending > 0
	}

	protected _get_pending_label(__: this, n_pending: number): string {
		return _("Pending Packages: %s").format(n_pending)
	}
}
